python/lsst/ap/verify/metrics.py

```diff
@@ -91,6 +93,15 @@ def __call__(self, _parser, _namespace, _values, option_string=None):
         warnings.warn(message, category=FutureWarning)
 
 
+class _OptionalFormatDict(collections.UserDict):
+    """A dictionary that, when used with a formatter, preserves unknown keys.
```

Comment:
I'm not understanding why this OptionalFormatDict is necessary here, can you please explain?

Reply:
Most of the string substitution (in particular, `{dataId}`) is done by `lsst.verify.gen2tasks.MetricsControllerTask`, so `ap_verify` shouldn't mess with it -- it should only substitute `{output}`, which is specific to `ap_verify`. But `str.format` raises an exception if it finds a replacement field it doesn't know how to fill in. So this class is basically a hack :pick: that makes `str.format_map` think it's replacing all the fields, but the replacements other than `{output}` are no-ops.

I'm not sure if that answers your question.

Reply:
That at least makes sense. It still appears convoluted to me, but I don't have an alternative to suggest. Can you please add a sentence or two to the docstring which explains the existence of this class (similar to what you commented here)?